src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,714 @@
+    }
+
+    /// @notice Set the Zap contract address.
+    function setZap(address _zap) external onlyOwner {
+        zap = _zap;
+
+        emit ZapUpdated(_zap);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Performance Fee Functions
+    ////////////////////////////////////////////////////
+
+    /// @notice Owner sets the performance fee on increased assets
+    /// @param _fee The performance fee measured in basis points (1/100th of a percent)
+    /// 10,000 = 100% performance fee
+    /// 500 = 5% performance fee
+    /// The max allowed performance fee is 50% (5000)
+    function setFee(uint256 _fee) external onlyOwner {
+        _setFee(_fee);
+    }
+
+    /// @notice Owner sets the account/contract that receives the performance fee
+    function setFeeCollector(address _feeCollector) external onlyOwner {
+        _setFeeCollector(_feeCollector);
+    }
+
+    function _setFee(uint256 _fee) internal {
+        require(_fee <= FEE_SCALE / 2, "ARM: fee too high");
+
+        // Collect any performance fees up to this point using the old fee
+        collectFees();
+
+        fee = SafeCast.toUint16(_fee);
+
+        emit FeeUpdated(_fee);
+    }
+
+    function _setFeeCollector(address _feeCollector) internal {
+        require(_feeCollector != address(0), "ARM: invalid fee collector");
+
+        feeCollector = _feeCollector;
+
+        emit FeeCollectorUpdated(_feeCollector);
+    }
+
+    /// @notice Transfer accrued performance fees to the fee collector
+    /// This requires enough liquidity assets (WETH) in the ARM that are not reserved
+    /// for the withdrawal queue to cover the accrued fees.
+    function collectFees() public returns (uint256 fees) {
+        uint256 newAvailableAssets;
+        // Accrue any performance fees up to this point
+        (fees, newAvailableAssets) = _feesAccrued();
+
+        if (fee == 0) return 0;
+
+        // Check there is enough liquidity assets (WETH) that are not reserved for the withdrawal queue
+        // to cover the fee being collected.
+        _requireLiquidityAvailable(fees);
+        // _requireLiquidityAvailable() is optimized for swaps so will not revert if there are no outstanding withdrawals.
+        // We need to check there is enough liquidity assets to cover the fees being collect from this ARM contract.
+        // We could try the transfer and let it revert if there are not enough assets, but there is no error message with
+        // a failed WETH transfer so we spend the extra gas to check and give a meaningful error message.
+        require(fees <= IERC20(liquidityAsset).balanceOf(address(this)), "ARM: insufficient liquidity");
+
+        IERC20(liquidityAsset).transfer(feeCollector, fees);
+
+        // Save the new available assets back to storage less the collected fees.
+        lastAvailableAssets = SafeCast.toInt128(SafeCast.toInt256(newAvailableAssets) - SafeCast.toInt256(fees));
```

Comment:
⚪️ We can move the `lastAvailableAssets =` up to before the transfer. This way we do computation and stage changes before transfers.

Reply:
Fix proposed in this commit [6d83336](https://github.com/OriginProtocol/arm-oeth/pull/33/commits/6d8333682b3b56076f19e3c537f64958a91f96bf).